velox/python/expression/expression.cpp

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <pybind11/operators.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "velox/python/expression/PyExpression.h"
+#include "velox/type/Variant.h"
+
+namespace py = pybind11;
+namespace {
+
+using namespace facebook;
+
+/// Converts a given Python object into a Velox constant expression. Throws
+/// TypeError if it cannot convert.
+velox::py::PyExpr toConstantExpr(const py::object& arg) {
+  // Bool (True|False).
+  if (py::isinstance<py::bool_>(arg)) {
+    return velox::py::PyExpr::createConstant(
+        velox::BOOLEAN(), velox::Variant(arg.cast<bool>()));
+  }
+  // None.
+  else if (py::isinstance<py::none>(arg)) {
+    return velox::py::PyExpr::createConstant(
+        velox::UNKNOWN(), velox::Variant::null(velox::TypeKind::UNKNOWN));
+  }
+  // Integers. For now we assume always bigint (64 bits).
+  else if (py::isinstance<py::int_>(arg)) {
+    return velox::py::PyExpr::createConstant(
+        velox::BIGINT(), velox::Variant(arg.cast<int64_t>()));
+  }
+  // Floats. Assume always double (64 bits) for now.
+  else if (py::isinstance<py::float_>(arg)) {
+    return velox::py::PyExpr::createConstant(
+        velox::DOUBLE(), velox::Variant(arg.cast<double>()));
+  }
+  // Strings.
+  else if (py::isinstance<py::str>(arg)) {
+    return velox::py::PyExpr::createConstant(
+        velox::VARCHAR(), velox::Variant(arg.cast<std::string>()));
+  }
+  // TODO: Support container types (arrays/list, maps, and structs).
+  throw py::type_error(
```

Comment:
FYI,  You can reuse/copy the code here (https://github.com/facebookincubator/velox/blob/main/velox/python/legacy/complex.cpp#L160) to add support for conversion of complex python types to velox vectors. It supports any kind of array right now, but is straightforward to add map/row support.